src/MotionHandler.moon

```diff
@@ -171,21 +171,55 @@ class MotionHandler
 				else
 					newText = line\detokenizeTransformsCopy timeDelta
 
-				newText = newText\gsub "\\fade(%b())", ( fade ) ->
-					a1, a2, a3, t1, t2, t3, t4 = fade\match("(%d+),(%d+),(%d+),(%d+),(%d+),(%d+),(%d+)")
-					if t1 == nil
-						message = "There is a malformed \\fade you must fix.\n\\fade requires 7 integer arguments.\nLine: #{.number}, tag: \\fade(#{fade})."
-						if fade\match("(%d+),(%d+)")
-							message ..= "\nPerhaps you meant to use \\fad."
-						log.windowError message
-
-					t1, t2, t3, t4 = tonumber( t1 ), tonumber( t2 ), tonumber( t3 ), tonumber( t4 )
-					-- beautiful.
-					t1 -= timeDelta
-					t2 -= timeDelta
-					t3 -= timeDelta
-					t4 -= timeDelta
-					("\\fade(%s,%s,%s,%d,%d,%d,%d)")\format a1, a2, a3, t1, t2, t3, t4
+				if @options.main.killTrans
+				-- multiplies every alpha tag by a scaling factor determined by the fade envelope
+
+					-- if there's a better way to do this I would love to know
+					a1, a2, a3, t1, t2, t3, t4 = 0, 0, 0, 0, 0, 0, 0
+					line\runCallbackOnFirstOverride ( tagBlock ) =>
+						return tagBlock\gsub "\\fade%((%d+),(%d+),(%d+),(%d+),(%d+),(%d+),(%d+)%)", ( a, b, c, d, e, f, g ) ->
+							a1, a2, a3, t1, t2, t3, t4 = tonumber a, tonumber b, tonumber c, tonumber d, tonumber e, tonumber f, tonumber g
+							return "",
+							1
+
+					-- Piecewise function done with logicals :)
+					fadeFactor = (
+						-- x < t1:
+						(x < t1) and 				a1 or 
+						-- t1 <= x < t2:
+						(t1 <= x) and (x < t2) and 	a1 + (a2 - a1) * (x - t1) / (t2 - t1) or
+						-- t2 <= x < t3
+						(t2 <= x) and (x < t3) and 	a2 or
+						-- t3 <= x < t4
+						(t3 <= x) and (x < t4) and 	a2 + (a3 - a2) * (x - t4) / (t4 - t3) or
+						-- x >= t4:
+						(x >= t4) and 				a3
+						)
+
+					fadeFactor = (255 - fadeFactor) / 255
+					-- factor between 0 and 1 representing opacity of fade line
+
+					line\runCallbackOnOverrides ( tagBlock ) =>
+						return tagBlock\gsub "(\\[1234]?a[lpha]-)&H(%x%x)&", ( alpha, value ) ->
+							value = Math.round( 255 - ( fadeFactor * (255 - tonumber value) ) )
+							return alpha .. "&H%x&"\format value)
+
```

Comment:
I would run all of the above work in the `fade` gsub callback, because currently it will run even if no `fade` exists in the line. It also saves you from having to worry about extracting the fade parameters into the upvalues.

I'm all about the arrow antipattern.